src/java/org/apache/cassandra/db/compaction/CompactionTask.java

```diff
@@ -155,114 +158,124 @@ public boolean apply(SSTableReader sstable)
-                                    lastCheckObsoletion = System.nanoTime();
-                                }
+                                controller.maybeRefreshOverlaps();
+                                lastCheckObsoletion = System.nanoTime();
                             }
                         }
+                    }
 
-                        readyToFinish = true;
-                        newSStables = writer.finish();
-                    } catch (Exception e) {
-                        CompactionException exception = new CompactionException(taskIdLoggerMsg, ssTableLoggerMsg.toString(), e);
-                        if (readyToFinish && e.getSuppressed() != null && e.getSuppressed().length != 0)
-                        {
-                            abortFailed = true;
-                            logger.warn("CompactionAwareWriter failed to close correctly for {}/{}. This compaction won't be removed from " +
-                                            "system.compactions_in_progress to ensure sstable cleanup on startup proceeds correctly in case some " +
-                                            "compaction-product sstables are marked final while others remain tmp",
-                                    cfs.keyspace.getName(), cfs.name, e);
-                        }
-                        throw exception;
+                    readyToFinish = true;
+                    newSStables = writer.finish();
+                } catch (Exception e) {
+                    CompactionException exception = new CompactionException(taskIdLoggerMsg, ssTableLoggerMsg.toString(), e);
+                    if (readyToFinish && e.getSuppressed() != null && e.getSuppressed().length != 0)
+                    {
+                        abortFailed = true;
+                        logger.warn("CompactionAwareWriter failed to close correctly for {}/{}. This compaction won't be removed from " +
+                                        "system.compactions_in_progress to ensure sstable cleanup on startup proceeds correctly in case some " +
+                                        "compaction-product sstables are marked final while others remain tmp",
+                                cfs.keyspace.getName(), cfs.name, e);
                     }
+                    throw exception;
                 }
             }
-            finally
-            {
-                Directories.removeExpectedSpaceUsedByCompaction(expectedWriteSize, CONSIDER_CONCURRENT_COMPACTIONS);
-                if (taskId != null && (!abortFailed))
-                    SystemKeyspace.finishCompaction(taskId);
-
-                if (collector != null && ci != null)
-                    collector.finishCompaction(ci);
+        }
+        finally
+        {
+            if (!readyToFinish) {
+                // TODO(wdey): refactor all of the trys
+                try (Refs closedRefs = refs; CompactionController closedController = controller) {}
             }
```

Comment:
this is unfortunate; we should do something like https://github.com/palantir/cassandra/pull/549#discussion_r1794454098 as a follow up